cmd/cluster/resize/infra_node.go

```diff
@@ -0,0 +1,308 @@
+			}
+		}
+
+		switch {
+		case readyNodes >= int(*originalMp.Spec.Replicas)*2:
+			return true, nil
+		default:
+			log.Printf("found %d infra nodes reporting Ready, continuing to wait", readyNodes)
+			return false, nil
+		}
+	}); err != nil {
+		return err
+	}
+
+	// Delete temp machinepool
+	log.Printf("deleting temporary machinepool %s, with machine type %s", tempMp.Name, tempMp.Spec.Platform.AWS.InstanceType)
+	if err := r.hiveAdmin.Delete(ctx, tempMp); err != nil {
+		return err
+	}
+
+	// Wait for temporary machinepool to delete
+	if err := wait.PollImmediate(twentySecondIncrement, twentyMinuteTimeout, func() (bool, error) {
+		mp := &hivev1.MachinePool{}
+		err := r.hive.Get(ctx, client.ObjectKey{Namespace: tempMp.Namespace, Name: tempMp.Name}, mp)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, err
+		}
+
+		log.Printf("temporary machinepool %s/%s still exists, continuing to wait", tempMp.Namespace, tempMp.Name)
+		return false, nil
+	}); err != nil {
+		return err
+	}
+
+	// Wait for infra node count to return to normal
+	log.Printf("waiting for infra node count to return to: %d", int(*originalMp.Spec.Replicas))
+	if err := wait.PollImmediate(twentySecondIncrement, twentyMinuteTimeout, func() (bool, error) {
+		nodes := &corev1.NodeList{}
+		selector, err := labels.Parse("node-role.kubernetes.io/infra=")
+		if err != nil {
+			return false, err
+		}
+
+		if err := r.client.List(ctx, nodes, &client.ListOptions{LabelSelector: selector}); err != nil {
+			return false, err
+		}
+
+		switch len(nodes.Items) {
+		case int(*originalMp.Spec.Replicas):
+			log.Printf("found %d infra nodes, infra resize complete", len(nodes.Items))
+			return true, nil
+		default:
+			log.Printf("found %d infra nodes, continuing to wait", len(nodes.Items))
+			return false, nil
+		}
+	}); err != nil {
+		return err
+	}
+
+	log.Printf("[REMINDER] follow the cleanup tasks in https://github.com/openshift/ops-sop/blob/master/v4/howto/resize-infras-workers.md#sending-the-all-clear")
+	return nil
+}
+
+func (r *Resize) getInfraMachinePool(ctx context.Context) (*hivev1.MachinePool, error) {
+	ns := &corev1.NamespaceList{}
+	selector, err := labels.Parse(fmt.Sprintf("api.openshift.com/id=%s", r.clusterId))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.hive.List(ctx, ns, &client.ListOptions{LabelSelector: selector, Limit: 1}); err != nil {
+		return nil, err
+	}
+	if len(ns.Items) == 0 {
+		return nil, fmt.Errorf("found 0 namespaces with tag: api.openshift.com/id=%s", r.clusterId)
+	}
+
+	log.Printf("found namespace: %s", ns.Items[0].Name)
+
+	mpList := &hivev1.MachinePoolList{}
+	if err := r.hive.List(ctx, mpList, &client.ListOptions{Namespace: ns.Items[0].Name}); err != nil {
+		return nil, err
+	}
+
+	for _, mp := range mpList.Items {
+		mp := mp
+		if mp.Spec.Name == "infra" {
+			log.Printf("found machinepool %s, with machine type %s", mp.Name, mp.Spec.Platform.AWS.InstanceType)
+			return &mp, nil
```

Comment:
**Question:** Is it possible that a cluster has two infra machinepools? E.g. because of a missing cleanup from a previous resize?

Reply:
Kind of - so it's not possible that a cluster has two machinepools with the name "infra" due to name uniqueness within the same namespace, but it is possible that a cluster has an "infra" and "infra2" if a previous cleanup went wrong.

I was thinking that this is up to the SRE to manually verify that a cluster is in a "reasonable" state